# The Android App Market on Google Play
Load, clean, and visualize scraped Google Play Store data to gain insights into the Android app market.



## Table of Contents:-

1.  Google Play Store apps and reviews

2.	Data cleaning

3.	Correcting data types

4.	Dealing with missing numerical values

5.	Exploring app categories

6.	Distribution of app ratings

7.	Size and price of an app

8.	Relation between app category and app price

9.	Filter out "junk" apps

10.	Popularity of paid apps vs free apps

11.	Sentiment analysis of user reviews



## 1. Google Play Store apps and reviews
<p>Mobile apps are everywhere. They are easy to create and can be lucrative. Because of these two factors, more and more apps are being developed. In this notebook, we will do a comprehensive analysis of the Android app market by comparing over ten thousand apps in Google Play across different categories. We'll look for insights in the data to devise strategies to drive growth and retention.</p>
<p><img src="https://assets.datacamp.com/production/project_619/img/google_play_store.png" alt="Google Play logo"></p>
<p>Let's take a look at the data, which consists of two files:</p>
<ul>
<li><code>apps.csv</code>: contains all the details of the applications on Google Play. There are 13 features that describe a given app.</li>
<li><code>user_reviews.csv</code>: contains 100 reviews for each app, <a href="https://www.androidpolice.com/2019/01/21/google-play-stores-redesigned-ratings-and-reviews-section-lets-you-easily-filter-by-star-rating/">most helpful first</a>. The text in each review has been pre-processed and attributed with three new features: Sentiment (Positive, Negative or Neutral), Sentiment Polarity and Sentiment Subjectivity.</li>
</ul>

In [307]:
# Read in dataset
import pandas as pd
apps_with_duplicates = pd.read_csv('datasets/apps.csv')


# Drop duplicates from apps_with_duplicates
apps = apps_with_duplicates.drop_duplicates()


print('Total number of apps in the dataset = ',len(apps))

# Have a look at a random sample of 5 rows
apps.sample(5)


Total number of apps in the dataset =  9659


Unnamed: 0                                      App          Category  \
9413       10588                     Florida Travel Guide  TRAVEL_AND_LOCAL   
5696        6727                                BS-Mobile     COMMUNICATION   
7672        8789            Dr.Android Repair Master 2017      PRODUCTIVITY   
9073       10243             Barcelona News - Sportfusion            SPORTS   
7561        8668  Adobe Fill & Sign: Easy PDF Form Filler      PRODUCTIVITY   

      Rating  Reviews  Size    Installs  Type Price Content Rating  \
9413     3.8       11  86.0      1,000+  Free     0       Everyone   
5696     5.0        1   0.7         50+  Free     0       Everyone   
7672     3.8     1205   4.3    100,000+  Free     0       Everyone   
9073     4.4      382  20.0     10,000+  Free     0       Everyone   
7561     4.3    11402   5.3  1,000,000+  Free     0       Everyone   

              Genres       Last Updated Current Ver Android Ver  
9413  Travel & Local      July 11, 2018       6.8.2  4.1 and up  
5696   Communication  September 3, 2015         0.1  2.3 and up  
7672    Productivity      June 10, 2018           2  4.0 and up  
9073          Sports      March 6, 2018       3.953  4.1 and up  
7561    Productivity   November 7, 2017       1.3.0  5.0 and up

In [308]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9659 entries, 0 to 9658
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      9659 non-null   int64  
 1   App             9659 non-null   object 
 2   Category        9659 non-null   object 
 3   Rating          8196 non-null   float64
 4   Reviews         9659 non-null   int64  
 5   Size            8432 non-null   float64
 6   Installs        9659 non-null   object 
 7   Type            9659 non-null   object 
 8   Price           9659 non-null   object 
 9   Content Rating  9659 non-null   object 
 10  Genres          9659 non-null   object 
 11  Last Updated    9659 non-null   object 
 12  Current Ver     9651 non-null   object 
 13  Android Ver     9657 non-null   object 
dtypes: float64(2), int64(2), object(10)
memory usage: 1.1+ MB


## 2. Data cleaning
<p>Data cleaning is one of the most essential subtask any data science project. Although it can be a very tedious process, it's worth should never be undermined.</p>
<p>By looking at a random sample of the dataset rows (from the above task), we observe that some entries in the columns like <code>Installs</code> and <code>Price</code> have a few special characters (<code>+</code> <code>,</code> <code>$</code>) due to the way the numbers have been represented. This prevents the columns from being purely numeric, making it difficult to use them in subsequent future mathematical calculations. Ideally, as their names suggest, we would want these columns to contain only digits from [0-9].</p>
<p>Hence, we now proceed to clean our data. Specifically, the special characters <code>,</code> and <code>+</code> present in <code>Installs</code> column and <code>$</code> present in <code>Price</code> column need to be removed.</p>
<p>It is also always a good practice to print a summary of your dataframe after completing data cleaning. We will use the <code>info()</code> method to acheive this.</p>

In [309]:
# List of characters to remove
chars_to_remove = ["+",",","$"]
# List of column names to clean
cols_to_clean = ['Installs','Price']

# Loop for each column in cols_to_clean
for col in cols_to_clean:
    # Loop for each char in chars_to_remove
    for char in chars_to_remove:
        # Replace the character with an empty string
        apps[col] = apps[col].apply(lambda x: x.replace(char,''))
        
# Print a summary of the apps dataframe
apps.sample(5)


Unnamed: 0                                                App  \
164          164                                       Ebook Reader   
4850        5835  25 Mins Ako ay may lobo Etc Pinoy Kid Song Off...   
6096        7140                                     Cádiz CB Gades   
5603        6629                                 BQ-መጽሐፍ ቅዱሳዊ ጥያቄዎች   
1276        1564                             Easy Hair Style Design   

                 Category  Rating  Reviews  Size Installs  Type Price  \
164   BOOKS_AND_REFERENCE     4.1    85842  37.0  5000000  Free     0   
4850               FAMILY     NaN        0  42.0       10  Free     0   
6096               SPORTS     NaN        1   3.8       10  Free     0   
5603                 GAME     4.7      191   7.2     5000  Free     0   
1276            LIFESTYLE     4.3      601   5.1   100000  Free     0   

     Content Rating             Genres       Last Updated        Current Ver  \
164        Everyone  Books & Reference      June 25, 2018              5.0.6   
4850     Mature 17+      Entertainment      July 29, 2018                  3   
6096       Everyone             Sports     August 5, 2018  opción de cerrar   
5603       Everyone             Trivia      July 31, 2018              4.1.2   
1276       Everyone          Lifestyle  December 20, 2017                  1   

     Android Ver  
164   4.0 and up  
4850  2.3 and up  
6096  4.1 and up  
5603  4.1 and up  
1276  2.3 and up

## 3. Correcting data types
<p>From the previous task we noticed that <code>Installs</code> and <code>Price</code> were categorized as <code>object</code> data type (and not <code>int</code> or <code>float</code>) as we would like. This is because these two columns originally had mixed input types: digits and special characters. To know more about Pandas data types, read <a href="https://datacarpentry.org/python-ecology-lesson/04-data-types-and-format/">this</a>.</p>
<p>The four features that we will be working with most frequently henceforth are <code>Installs</code>, <code>Size</code>, <code>Rating</code> and <code>Price</code>. While <code>Size</code> and <code>Rating</code> are both <code>float</code> (i.e. purely numerical data types), we still need to work on <code>Installs</code> and <code>Price</code> to make them numeric.</p>

In [310]:
import numpy as np

# Convert Installs to float data type
apps['Installs'] = apps['Installs'].astype('float')

# Convert Price to float data type
apps['Price'] = apps['Price'].astype('float')

# Checking dtypes of the apps dataframe
print(apps.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9659 entries, 0 to 9658
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      9659 non-null   int64  
 1   App             9659 non-null   object 
 2   Category        9659 non-null   object 
 3   Rating          8196 non-null   float64
 4   Reviews         9659 non-null   int64  
 5   Size            8432 non-null   float64
 6   Installs        9659 non-null   float64
 7   Type            9659 non-null   object 
 8   Price           9659 non-null   float64
 9   Content Rating  9659 non-null   object 
 10  Genres          9659 non-null   object 
 11  Last Updated    9659 non-null   object 
 12  Current Ver     9651 non-null   object 
 13  Android Ver     9657 non-null   object 
dtypes: float64(4), int64(2), object(8)
memory usage: 1.1+ MB
None


## 4. Dealing with missing numerical values
The following commands help us to deal with missing numerical values.


`apps.isnull()`

The above command checks whether each cell in a dataframe contains missing values or not. If the cell contains missing value, it returns True otherwise it returns False. 


`apps.isnull.sum()`

The above command returns the total number of missing values in each column in the dataset.



**isna()** and **notna()** functions to detect ‘NA’ values


Pandas provides isna() and notna() functions to detect ‘NA’ values. These are also methods on Series and DataFrame objects.


Examples of isna() and notna() commands


We can detect ‘NA’ values in the dataframe with the following command:-


`apps.isna()`



We can detect ‘NA’ values in a particular column in the dataframe


`pd.isna(apps[‘col_name’])`


`apps[‘col_name’].notna()`


So, we can check the number of missing values in each column in the dataset as follows:-


In [311]:
apps.isnull().sum()

Unnamed: 0           0
App                  0
Category             0
Rating            1463
Reviews              0
Size              1227
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

We can see that there are lots of missing values in the dataset.

### Fill missing values with a test statistic


In this method, we fill the missing values with a test statistic like mean, median or mode of the particular feature the missing value belongs to. One can also specify a forward-fill or back-fill to propagate the next values backward or previous value forward.



### Filling missing values with a test statistic like mean


`mean = apps['col_name'].mean()`


`apps['col_name'].fillna(value = median, inplace = True )`





We can calculate the missing variables and use it to fill the missing values as follows:-

In [312]:
mean_rating= apps['Rating'].mean()
apps['Rating'].fillna(mean_rating, inplace=True)
apps['Rating']=round(apps['Rating'],2)

In [313]:
mean_rating= apps['Size'].mean()
apps['Size'].fillna(mean_rating, inplace=True)
apps['Size']=round(apps['Size'],2)

In [314]:
apps.sample(5)


Unnamed: 0                                   App       Category  Rating  \
8100        9220            Electrician Calculator Pro          TOOLS     3.9   
3260        4117        Inputting Plus: Ctrl + Z/F/C/V   PRODUCTIVITY     4.2   
2606        3335                     WiFi Overview 360          TOOLS     4.1   
4018        4961  Ad Blocker Turbo - Adblocker Browser  COMMUNICATION     3.9   
5059        6052        Be My Eyes - Helping the blind      LIFESTYLE     4.9   

      Reviews  Size   Installs  Type  Price Content Rating         Genres  \
8100       47   0.3     1000.0  Paid  14.99       Everyone          Tools   
3260     1117   3.0    50000.0  Free   0.00       Everyone   Productivity   
2606    15693  18.0  1000000.0  Free   0.00       Everyone          Tools   
4018      107  44.0    10000.0  Free   0.00       Everyone  Communication   
5059     8418  21.0   500000.0  Free   0.00       Everyone      Lifestyle   

          Last Updated Current Ver Android Ver  
8100      June 3, 2014         3.1  1.6 and up  
3260  October 18, 2017       1.0.6  5.0 and up  
2606    August 3, 2018     4.32.07  4.2 and up  
4018  January 29, 2018       1.0.5  4.1 and up  
5059     July 11, 2018       1.2.9  4.4 and up

In [315]:
apps.isnull().sum()

Unnamed: 0        0
App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       8
Android Ver       2
dtype: int64

We can see that there are no missing values in the dataframe. 



## 5. Exploring app categories
<p>With more than 1 billion active users in 190 countries around the world, Google Play continues to be an important distribution platform to build a global audience. For businesses to get their apps in front of users, it's important to make them more quickly and easily discoverable on Google Play. To improve the overall search experience, Google has introduced the concept of grouping apps into categories.</p>
<p>This brings us to the following questions:</p>
<ul>
<li>Which category has the highest share of (active) apps in the market? </li>
<li>Is any specific category dominating the market?</li>
<li>Which categories have the fewest number of apps?</li>
</ul>


In [316]:
pip install plotly

You should consider upgrading via the 'c:\Users\User\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [317]:
import plotly
plotly.offline.init_notebook_mode(connected=True)
import plotly.graph_objs as go

# Print the total number of unique categories
num_categories =len(apps['Category'].unique())
print('Number of categories = ',num_categories)

# Count the number of apps in each 'Category'. 
num_apps_in_category = apps['Category'].value_counts()
# Sort num_apps_in_category in descending order based on the count of apps in each category
sorted_num_apps_in_category = num_apps_in_category.sort_values(ascending=False)

data = [go.Bar(
        x = num_apps_in_category.index, # index = category name
        y = num_apps_in_category.values, # value = count
       
)]
layout = dict(
    title = "The distribution of apps across different categories.",
    xaxis = dict(
        title="Category name",
        
        ),
    yaxis = dict(
        title="Number of apps in category ",
        ),
)

fig = dict(data=data, layout=layout)


plotly.offline.iplot(fig)

Number of categories =  33


<p>We  see that there are <code>33</code> unique app categories present in our dataset. <code><em>Family</em></code> and <code><em>Game</em></code> apps have the highest market prevalence. Interestingly, <code><em>Tools</em></code>, <code><em>Business</em></code> and <code><em>Medical</em></code> apps are also at the top.</p>
<p> We see that also <code><em>Art and Design</em></code>, <code><em>Comics</em></code> and <code><em>Beauty</em></code> categories  have the fewest number of apps.</p>

## 6. Distribution of app ratings
<p>After having witnessed the market share for each category of apps, let's see how all these apps perform on an average. App ratings (on a scale of 1 to 5) impact the discoverability, conversion of apps as well as the company's overall brand image. Ratings are a key performance indicator of an app.</p>


In [318]:
# Average rating of apps
avg_app_rating = apps['Rating'].mean()
print('Average app rating = ', avg_app_rating)

# Distribution of apps according to their ratings
data = [go.Histogram(
        x = apps['Rating']
)]

# Vertical dashed line to indicate the average app rating
layout = {'shapes': [{
              
              'type' :'line',
              'x0': avg_app_rating,
              'y0': 0,
              'x1': avg_app_rating,
              'y1': 1000,
              'line': { 'dash': 'dashdot'}
          }],
        'title':'The average app rating',
        'xaxis': { 'title': 'Average of rating'},
        'yaxis': { 'title': 'Number of apps'}
        }
          

plotly.offline.iplot({'data': data, 'layout': layout})

Average app rating =  4.172751837664355


<p>From our research, we found that the average volume of ratings across all app categories is <code>4.17</code>. The histogram plot is skewed to the left indicating that the majority of the apps are highly rated with only a few exceptions in the low-rated apps.</p>

## 7. Size and price of an app
<p>Let's now examine app size and app price. For size, if the mobile app is too large, it may be difficult and/or expensive for users to download. Lengthy download times could turn users off before they even experience your mobile app. Plus, each user's device has a finite amount of disk space. For price, some users expect their apps to be free or inexpensive. These problems compound if the developing world is part of your target market; especially due to internet speeds, earning power and exchange rates.</p>
<p>How can we effectively come up with strategies to size and price our app?</p>
<ul>
<li>Does the size of an app affect its rating? </li>
<li>Do users really care about system-heavy apps or do they prefer light-weighted apps? </li>
<li>Does the price of an app affect its rating? </li>
<li>Do users always prefer free apps over paid apps?</li>
</ul>


In [319]:
pip install matplotlib

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\User\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [320]:
pip install seaborn

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'c:\Users\User\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [321]:
%matplotlib inline
import seaborn as sns
sns.set_style("darkgrid")
import warnings
warnings.filterwarnings("ignore")


# Subset for categories with at least 250 apps
large_categories = apps.groupby(apps['Category']).filter(lambda x: len(x) >= 250)

# Plot size vs. rating
plt1 = sns.jointplot(x = large_categories['Size'], y = large_categories['Rating'])

# Select apps whose 'Type' is 'Paid'
paid_apps = apps[apps['Type']=='Paid']

# Plot price vs. rating
plt2 = sns.jointplot(x = paid_apps['Price'], y = paid_apps['Rating'])

<p>We find that the majority of top rated apps (rating over 4) range from 2 MB to 20 MB. We also find that the vast majority of apps price themselves under \$10.</p>

## 8. Relation between app category and app price
<p>So now comes the hard part. How are companies and developers supposed to make ends meet? What monetization strategies can companies use to maximize profit? The costs of apps are largely based on features, complexity, and platform.</p>
<p>There are many factors to consider when selecting the right pricing strategy for your mobile app. It is important to consider the willingness of your customer to pay for your app. A wrong price could break the deal before the download even happens. Potential customers could be turned off by what they perceive to be a shocking cost, or they might delete an app they’ve downloaded after receiving too many ads or simply not getting their money's worth.</p>
<p>Different categories demand different price ranges. Some apps that are simple and used daily, like the calculator app, should probably be kept free. However, it would make sense to charge for a highly-specialized medical app that diagnoses diabetic patients. Below, we see that <em>Medical and Family</em> apps are the most expensive. Some medical apps extend even up to \$80! All game apps are reasonably priced below \$20.</p>

In [ ]:
import matplotlib.pyplot as plt
fig, ax = plt.subplots()
fig.set_size_inches(15, 8)

# Select a few popular app categories
popular_app_cats = apps[apps.Category.isin(['GAME', 'FAMILY', 'PHOTOGRAPHY',
                                            'MEDICAL', 'TOOLS', 'FINANCE',
                                            'LIFESTYLE','BUSINESS'])]

# Examine the price trend by plotting Price vs Category
ax = sns.stripplot(x = popular_app_cats['Price'], y = popular_app_cats['Category'], jitter=True, linewidth=1)
ax.set_title('App pricing trend across categories')

# Apps whose Price is greater than 200
apps_above_200 = popular_app_cats[popular_app_cats['Price']>200]
apps_above_200[['Category', 'App', 'Price']]


## 9. Filter out "junk" apps
<p>It looks like a bunch of the really expensive apps are "junk" apps. That is, apps that don't really have a purpose. Some app developer may create an app called <em>I Am Rich Premium</em> or <em>most expensive app (H)</em> just for a joke or to test their app development skills. Some developers even do this with malicious intent and try to make money by hoping people accidentally click purchase on their app in the store.</p>
<p>Let's filter out these junk apps and re-do our visualization.</p>

In [ ]:
# Select apps priced below $100
apps_under_100 = popular_app_cats[popular_app_cats['Price']<100]

fig, ax = plt.subplots()
fig.set_size_inches(15, 8)

# Examine price vs category with the authentic apps (apps_under_100)
ax = sns.stripplot(x = apps_under_100 ['Price'], y = apps_under_100 ['Category'], data = apps_under_100, jitter = True, linewidth = 1)
ax.set_title('App pricing trend across categories after filtering for junk apps')

## 10. Popularity of paid apps vs free apps
<p>For apps in the Play Store today, there are five types of pricing strategies: free, freemium, paid, paymium, and subscription. Let's focus on free and paid apps only. Some characteristics of free apps are:</p>
<ul>
<li>Free to download.</li>
<li>Main source of income often comes from advertisements.</li>
<li>Often created by companies that have other products and the app serves as an extension of those products.</li>
<li>Can serve as a tool for customer retention, communication, and customer service.</li>
</ul>
<p>Some characteristics of paid apps are:</p>
<ul>
<li>Users are asked to pay once for the app to download and use it.</li>
<li>The user can't really get a feel for the app before buying it.</li>
</ul>
<p>Are paid apps installed as much as free apps? It turns out that paid apps have a relatively lower number of installs than free apps, though the difference is not as stark as I would have expected!</p>

In [ ]:
trace0 = go.Box(
    # Data for paid apps
    y = apps[apps['Type'] == 'Paid']['Installs'],
    name = 'Paid'
)

trace1 = go.Box(
    # Data for free apps
    y = apps[apps['Type'] == 'Free']['Installs'],
    name = 'Free'
)

layout = go.Layout(
    title = "Number of downloads of paid apps vs. free apps",
    yaxis = dict(title = "Log number of downloads",
                type = 'log',
                autorange = True)
)

# Add trace0 and trace1 to a list for plotting
data = [trace0, trace1]
plotly.offline.iplot({'data': data, 'layout': layout})

## 11. Sentiment analysis of user reviews
<p>Mining user review data to determine how people feel about your product, brand, or service can be done using a technique called sentiment analysis. User reviews for apps can be analyzed to identify if the mood is positive, negative or neutral about that app. For example, positive words in an app review might include words such as 'amazing', 'friendly', 'good', 'great', and 'love'. Negative words might be words like 'malware', 'hate', 'problem', 'refund', and 'incompetent'.</p>


In [ ]:
# Load user_reviews.csv
reviews_df = pd.read_csv('datasets/user_reviews.csv')

# Join the two dataframes
merged_df = pd.merge(apps, reviews_df, on = "App")

# Drop NA values from Sentiment and Review columns
merged_df = merged_df.dropna(subset = ['Sentiment', 'Review'])

sns.set_style('ticks')
fig, ax = plt.subplots()
fig.set_size_inches(11, 8)

# User review sentiment polarity for paid vs. free apps
ax = sns.boxplot(x = 'Type', y = 'Sentiment_Polarity', data = merged_df)
ax.set_title('Sentiment Polarity Distribution')

<p>By plotting sentiment polarity scores of user reviews for paid and free apps, we observe that free apps receive a lot of harsh comments, as indicated by the outliers on the negative y-axis. Reviews for paid apps appear never to be extremely negative. This may indicate something about app quality, i.e., paid apps being of higher quality than free apps on average. The median polarity score for paid apps is a little higher than free apps, thereby syncing with our previous observation.</p>


<p>In this notebook, we analyzed over ten thousand apps from the Google Play Store. We can use our findings to inform our decisions should we ever wish to create an app ourselves.</p>